assets/js/admin/marketplace-suggestions.js

```diff
@@ -0,0 +1,433 @@
+
+			var utmUrl = addUTMParameters( context, url );
+			linkoutButton.setAttribute( 'href', utmUrl );
+			linkoutButton.setAttribute( 'target', 'blank' );
+			linkoutButton.textContent = text;
+
+			linkoutButton.onclick = function() {
+				window.wcTracks.recordEvent( 'marketplace_suggestion_clicked', {
+					suggestionSlug: slug
+				} );
+			};
+
+			if ( isButton ) {
+				linkoutButton.classList.add( 'button' );
+			} else {
+				linkoutButton.classList.add( 'linkout' );
+				var linkoutIcon = document.createElement( 'span' );
+				linkoutIcon.classList.add( 'dashicons', 'dashicons-external' );
+				linkoutButton.appendChild( linkoutIcon );
+			}
+
+			return linkoutButton;
+		}
+
+		// Render DOM element for suggestion icon image.
+		function renderSuggestionIcon( iconUrl ) {
+			if ( ! iconUrl ) {
+				return null;
+			}
+
+			var image = document.createElement( 'img' );
+			image.src = iconUrl;
+			image.classList.add( 'marketplace-suggestion-icon' );
+
+			return image;
+		}
+
+		// Render DOM elements for suggestion content.
+		function renderSuggestionContent( title, copy ) {
+			var container = document.createElement( 'div' );
+
+			container.classList.add( 'marketplace-suggestion-container-content' );
+
+			if ( title ) {
+				var titleHeading = document.createElement( 'h4' );
+				titleHeading.textContent = title;
+				container.appendChild( titleHeading );
+			}
+
+			if ( copy ) {
+				var body = document.createElement( 'p' );
+				body.textContent = copy;
+				container.appendChild( body );
+			}
+
+			return container;
+		}
+
+		// Render DOM elements for suggestion call-to-action – button or link with dismiss 'x'.
+		function renderSuggestionCTA( context, slug, url, linkText, linkIsButton, allowDismiss ) {
+			var container = document.createElement( 'div' );
+
+			if ( ! linkText ) {
+				linkText = marketplace_suggestions.i18n_marketplace_suggestions_default_cta;
+			}
+
+			container.classList.add( 'marketplace-suggestion-container-cta' );
+			if ( url && linkText ) {
+				var linkoutElement = renderLinkout( context, slug, url, linkText, linkIsButton );
+				container.appendChild( linkoutElement );
+			}
+
+			if ( allowDismiss ) {
+				container.appendChild( renderDismissButton( context, slug ) );
+			}
+
+			return container;
+		}
+
+		// Render a "table banner" style suggestion.
+		// These are used in admin lists, e.g. products list.
+		function renderTableBanner( context, slug, iconUrl, title, copy, url, buttonText, allowDismiss ) {
+			if ( ! title || ! url ) {
+				return;
+			}
+
+			var row = document.createElement( 'tr' );
+			row.classList.add( 'marketplace-table-banner' );
+			row.classList.add( 'marketplace-suggestions-container' );
+			row.classList.add( 'showing-suggestion' );
+			row.dataset.marketplaceSuggestionsContext = 'products-list-inline';
+			row.dataset.suggestionSlug = slug;
+
+			var cell = document.createElement( 'td' );
+			cell.setAttribute( 'colspan', 10 );
```

Comment:
Testing this out with a backup from one of my "live" sites, and encountered an issue with this logic here. I have the Facebook extension installed which adds a column to the table, but other extensions do a similar thing, and users can also toggle columns on/off. So this is what the suggestion looks like to me:

![colspan](https://user-images.githubusercontent.com/22080/53844883-fb2b3f00-3f5b-11e9-8e49-776489077b58.png)

We could lean on jquery a bit here to determine the right number of columns that are actively displayed:

`$( 'table.wp-list-table.posts thead th:not(.hidden)' ).length`

Which would give us the correct number of th's that are not currently hidden. I _think_ that would work. Along those lines too, but perhaps a bit more edge-case, is when users toggle column visibility, things would still get wonky, even if adjusting the colspan accordingly. We'd have to listen for those events and respond accordingly. Not sure if that is a critical problem or not, but something to consider.

Reply:
Screen grab of toggling columns: https://drive.google.com/file/d/1IkapaDFhEFBGnB61D5kjXfpNv2Os9MNI/view?usp=sharing

Reply:
Thanks for finding that – will look at implementing the jQuery based fix.

Reply:
It's now calculating the `colspan` when the banner is initially displayed.

I've also added an event handler to catch when screen options change and recalculate it. There are probably some other cases we could catch – keeping it simple for now.

https://github.com/woocommerce/woocommerce/pull/22857/commits/f944da3eacaf72e381fb3a856dbe2e3f17612021